test/EFCore.PG.FunctionalTests/NpgsqlDatabaseCreatorTest.cs

```diff
@@ -1,397 +1,626 @@
+                        }
+                        else
+                        {
+                            Assert.True(context.Database.EnsureDeleted());
+                        }
+                    }
+                });
 
-        [Fact]
-        public async Task HasTablesAsync_returns_false_when_database_exists_but_has_no_tables()
-        {
-            await HasTables_returns_false_when_database_exists_but_has_no_tables_test(async: true);
-        }
+            Assert.Equal(ConnectionState.Closed, context.Database.GetDbConnection().State);
 
-        static async Task HasTables_returns_false_when_database_exists_but_has_no_tables_test(bool async)
-        {
-            using (var testDatabase = NpgsqlTestStore.CreateScratch(createDatabase: true))
-            {
-                var creator = GetDatabaseCreator(testDatabase);
+            Assert.False(async ? await creator.ExistsAsync() : creator.Exists());
 
-                Assert.False(async
-                    ? await ((TestDatabaseCreator)creator).HasTablesAsyncBase()
-                    : ((TestDatabaseCreator)creator).HasTablesBase());
-            }
+            Assert.Equal(ConnectionState.Closed, context.Database.GetDbConnection().State);
         }
 
-        [Fact]
-        public async Task HasTables_returns_true_when_database_exists_and_has_any_tables()
+        [ConditionalTheory]
+        [InlineData(true)]
+        [InlineData(false)]
+        private static async Task Noop_when_database_does_not_exist_test(bool async)
         {
-            await HasTables_returns_true_when_database_exists_and_has_any_tables_test(async: false);
-        }
+            using var testDatabase = NpgsqlTestStore.Create("NonExisting");
+            using var context = new BloggingContext(testDatabase);
+            var creator = GetDatabaseCreator(context);
 
-        [Fact]
-        public async Task HasTablesAsync_returns_true_when_database_exists_and_has_any_tables()
-        {
-            await HasTables_returns_true_when_database_exists_and_has_any_tables_test(async: true);
-        }
+            Assert.False(async ? await creator.ExistsAsync() : creator.Exists());
 
-        static async Task HasTables_returns_true_when_database_exists_and_has_any_tables_test(bool async)
-        {
-            using (var testDatabase = NpgsqlTestStore.CreateScratch(createDatabase: true))
+            if (async)
             {
-                await testDatabase.ExecuteNonQueryAsync("CREATE TABLE some_table (id SERIAL PRIMARY KEY)");
-
-                var creator = GetDatabaseCreator(testDatabase);
-
-                Assert.True(async ? await ((TestDatabaseCreator)creator).HasTablesAsyncBase() : ((TestDatabaseCreator)creator).HasTablesBase());
+                Assert.False(await creator.EnsureDeletedAsync());
+            }
+            else
+            {
+                Assert.False(creator.EnsureDeleted());
             }
```

Comment:
For example this can be rewritten as a three line statement:

```csharp
Assert.False(async
    ? await creator.EnsureDeletedAsync()
    : creator.EnsureDeleted());
```